### PR TITLE
Pe 8 provisiond schema conversion from xsd to internal json object

### PR DESCRIPTION
**Pull request by freddy61025**
https://issues.opennms.org/browse/PE-8

ref: https://github.com/OpenNMS/opennms/blob/PE-8-provisiond-schema-conversion-from-xsd-to-internal-json-object/features/config/service/impl/src/main/java/org/opennms/features/config/service/impl/ConfigurationManagerServiceImpl.java

https://issues.opennms.org/browse/PE-19

ref: https://github.com/OpenNMS/opennms/blob/PE-8-provisiond-schema-conversion-from-xsd-to-internal-json-object/features/config/service/impl/src/main/java/org/opennms/features/config/service/impl/ConfigurationManagerServiceImpl.java

https://issues.opennms.org/browse/PE-22

validation: https://github.com/OpenNMS/opennms/blob/PE-8-provisiond-schema-conversion-from-xsd-to-internal-json-object/features/config/dao/impl/src/main/java/org/opennms/features/config/dao/impl/JsonConfigStoreDaoImpl.java

### CODE REVIEW

**Review by patrick-schweizer — COMMENTED** [review]
I added some comments some of them super minor...

File: features/config/dao/api/src/main/java/org/opennms/features/config/dao/api/ServiceSchema.java [hunk 1]
```diff
@@ -29,25 +29,36 @@
...
 
 import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.ws.commons.schema.XmlSchemaCollection;
+import org.opennms.features.config.dao.api.util.XsdModelConverter;
 
 import java.util.Objects;
 
 public final class ServiceSchema {
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply features/config/dao/api/src/main/java/org/opennms/features/config/dao/api/ServiceSchema.java]: rename ServiceSchema > XmlValidationSchema for more easier understanding

File: features/config/dao/api/src/main/java/org/opennms/features/config/dao/api/ServiceSchema.java [hunk 2]
```diff
@@ -29,25 +29,36 @@
...
     private final XMLSchema xmlSchema;
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply features/config/dao/api/src/main/java/org/opennms/features/config/dao/api/ServiceSchema.java]: Created a ValidationSchema Interface for future  extension

File: features/config/dao/impl/src/main/java/org/opennms/features/config/dao/impl/JsonConfigStoreDaoImpl.java [hunk 3]
```diff
@@ -116,7 +116,6 @@ public void updateConfigSchema(ConfigSchema<?> configSchema) throws IOException
     @Override
     public Optional<ConfigData<JSONObject>> getConfigData(final String configName) throws IOException {
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply features/config/dao/impl/src/main/java/org/opennms/features/config/dao/impl/JsonConfigStoreDaoImpl.java:116]: Added more comment on ConfigStoreDao about that. getConfigData mainly design for CM internal use. External client should use getConfig(String, String)
    /** 
     * get configs data by configName
     * It gives the raw ConfigData object.
     * If you only want to get the specific config, you should use getConfig(String, String)
     * @see #getConfig(String, String)
     * @see ConfigData
     * @param configName
     * @return config object
     * @throws IOException
     */
    Optional<ConfigData<CONFIG_DATATYPE>> getConfigData(String configName) throws IOException;

File: features/config/service/impl/src/test/java/org/opennms/features/config/service/impl/ConfigurationManagerServiceImplTest.java [hunk 4]
```diff
@@ -64,59 +65,83 @@
...
+        String xmlStr = Files.readString(Path.of(xmlPath.getPath()));
+        Object configObject = configSchema.get().getConverter().xmlToJaxbObject(xmlStr);
+        configManagerService.registerConfiguration(CONFIG_NAME, CONFIG_ID, configObject);
+    }
+
+    @After
+    public void after() {
+        try {
+            configManagerService.unregisterSchema(CONFIG_NAME);
+        } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply features/config/service/impl/src/test/java/org/opennms/features/config/service/impl/ConfigurationManagerServiceImplTest.java]: removed

File: features/config/service/impl/src/test/java/org/opennms/features/config/service/impl/ConfigurationManagerServiceImplTest.java [hunk 5]
```diff
@@ -64,59 +65,83 @@
...
-        Assert.assertTrue("Cannot get XML config", xml.get().length() > 0);
-        Optional<ProvisiondConfiguration> entityFromDb = configManagerService.getConfiguration(SERVICE_NAME, CONFIG_ID,
-                ProvisiondConfiguration.class);
-        Assert.assertTrue("Cannot get XML config", entityFromDb.get().getImportThreads() == 11);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void testRegisterInvalidConfiguration() throws IOException {
+        ProvisiondConfiguration config = new ProvisiondConfiguration();
+        config.setImportThreads(-1L);
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply features/config/service/impl/src/test/java/org/opennms/features/config/service/impl/ConfigurationManagerServiceImplTest.java:112]: comment added

**Review by patrick-schweizer — APPROVED** [review]

**Pull request merged**